### PR TITLE
[@yaireo/tagify] Add types for react wrapper

### PR DESCRIPTION
**Pull request by blutorange**
Alright, so I took a look at the React wrapper and added the type declarations. Changes I made:

* Added types for React wrapper
* Adjusted a few docs
* Changed type of `tagTextProp` from `string` to `keyof T`, as one should specify a valid existing property of the tag data. When the type parameter T is set to the default `TagData`, `keyof T` is still `string`, since `TagData` allows everything.
* Added missing property `suggestedListItems` to the Tagify class (mentioned in the README). It is optional because it is possible to access a Tagify instance when that property is still unset, e.g. via the this context of the function in the `templates` setting.
* <strike>Omitted `MixedTags` component, as it's not actually exported from the bundled file currently (bug? yairEO/tagify#790)</strike>
* <strike>Omitted `children` prop for now because it is (a) deprecated according the docs, and (b) not working. The intended type is also unclear to me. yairEO/tagify#791</strike>
* <strike>Omitted `defaultValue` for now because React logs an error in dev mode (cannot set both value and defaultValue) and I'm not sure if the prop `defaultValue` is there to stay.</strike>
* <strike>Omitted `onEdit` for now because it is not working (only edit:input, edit:keydown etc. works). yairEO/tagify#792</strike>

---

### Remaining issues 

I also made a sample React project for myself so I could test the React wrapper and so I could see how it works. I did find a few issues, will probably open an issue over at the tagify repo... for now here are the issues I found as a note to myself for later... at the very least we should probably clear up the missing MixedTags export though

* <strike>React prop `showFilteredDropdown` in doc should be renamed to `showDropdown`</strike>
* <strike>Probably since the change to how `onChange` works, react (in dev mode) issues the below warning (as `onChange` was removed from the input / textarea).</strike>
> Warning: You provided a `value` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultValue`. Otherwise, set either `onChange` or `readOnly`.
* <strike>The `Tags` component also has the prop `defaultValue`, but since the `value` is always set, this results in even more errors:</strike>
> Warning: App contains an input of type text with both value and defaultValue props. Input elements must be either controlled or uncontrolled (specify either the value prop, or the defaultValue prop, but not both). Decide between using a controlled or uncontrolled input element and remove one of these props. More info: https://reactjs.org/link/controlled-components
* Settting `loading=true` and `showDropdown="..."` initially does not work, only when set later
* Setting `showDropdown=true` does not work. Also, I noticed that setting an empty string for `showDropdown` on the React component hides the dropdown, whereas `tagify.dropdown.show.call(tagify, "")` shows the dropdown with all available tags in the whitelist. Not sure if that is intentional though.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/yairEO/tagify#react
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by Brakebein — COMMENTED** [review]
Ok, I don't know much about React. I will just trust that everything React-related is fine and works.

I saw that the first CI test run failed:

>     Error: yaireo__tagify: Only index.d.ts may be listed explicitly in tsconfig's "files" entry.
>     Other d.ts files must either be referenced through index.d.ts, tests, or added to OTHER_FILES.txt.

The [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#other_filestxt) says:

> If a file is neither tested nor referenced in index.d.ts, add it to a file named OTHER_FILES.txt. This file is a list of other files that need to be included in the typings package, one file per line.

In order that the CI test succeeds, you removed `react.tagify.d.ts` from `tsconfig.json`. But I'm not sure if `OTHER_FILES.txt` is neccessary in order that the new d.ts file with additional entry point will be included in the final npm package.

Do you want to add the `OTHER_FILES.txt`? Or will we just look if the `react.tagify.d.ts` will be included in the final npm package, and if not making a new PR adding this txt file?

**Comment by blutorange** [comment]
> In order that the CI test succeeds, you removed react.tagify.d.ts from tsconfig.json. But I'm not sure if OTHER_FILES.txt is neccessary in order that the new d.ts file with additional entry point will be included in the final npm package.

> Do you want to add the OTHER_FILES.txt? Or will we just look if the react.tagify.d.ts will be included in the final npm package, and if not making a new PR adding this txt file?

Yeah, I had to try that out a little bit as well. Actually, I did add a `OTHER_FILES.txt` locally and ran the tests, this resulted in an error because the `react.tagify.d.ts` is already reachable via a test. The [README says this regarding adding new packages](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package):

> If you have .d.ts files besides index.d.ts, make sure that they are referenced either in index.d.ts or the tests.

So I think it should be OK the way it is now. If that really should not work, we can make another PR

**Comment by blutorange** [comment]
@Brakebein I would, however, like to wait a bit for the issues I opened at `@yaireo/tagify` (see first comment). The author seems pretty active

**Review by Brakebein — CHANGES_REQUESTED** [review]
You'll need to bump the version to `4.0.1` in the first line of `index.d.ts`. Otherwise it won't update the npm package.

File: types/yaireo__tagify/index.d.ts [hunk 1]
```diff
@@ -46,6 +46,11 @@ declare namespace Tagify {
     /**
      * Settings for the dropdown feature that can be configured via the
      * `dropdown` option of the settings that are passed to tagify.
+     * @template T Type of the tag data used by the tagify instance. Must
```
[the remark this answers is not in the document]
Comment (blutorange) [reply types/yaireo__tagify/index.d.ts:49]: I don't think so, but I think I'll just keep the long comment on the Tagify class and shorten the rest to one line.

**Comment by Brakebein** [comment]
> @Brakebein I would, however, like to wait a bit for the issues I opened at `@yaireo/tagify` (see first comment). The author seems pretty active

Ok, I will wait. Just say when you think you want to merge.

**Comment by blutorange** [comment]
> You'll need to bump the version to `4.0.1` in the first line of `index.d.ts`. Otherwise it won't update the npm package.

@Brakebein Are you sure about that? As far as I understand it, the patch version of the `@types` is for fixes to the typings package itself (patch version of the main npm package should not add any new features). The version in the `index.d.ts` is usually just major.minor (see e.g. types for base64-js, jquery, react etc.). So it should just increment the patch version of `@types` automatically during publish.

https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library

**Comment by Brakebein** [comment]
You are right, sorry. You just define major and minor version with respect to the original package. I just messed it up with "normal" packages where you need to increment the patch version by yourself.

**Review by Brakebein — APPROVED** [review]
Now, you can merge if you want, or I will wait for new commits.

**Comment by blutorange** [comment]
@Brakebein The issues over at yaireo/tagify have been resolved now, if you could take a last look then I think we're ready to merge when the CI is successful.

**Review by Brakebein — APPROVED** [review]
Ready to merge

**Review by rbuckton — APPROVED** [review]

**Comment by Brakebein** [comment]
Ready to merge

**Pull request merged**